components/brave_vpn/brave_vpn_connection_manager_win.cc

```diff
@@ -0,0 +1,121 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_vpn/brave_vpn_connection_manager_win.h"
+
+#include "base/logging.h"
+#include "base/notreached.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/post_task.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/brave_vpn/utils_win.h"
+
+// Most of Windows implementations are based on Brian Clifton
+// (brian@clifton.me)'s work (https://github.com/bsclifton/winvpntool).
+
+using brave_vpn::internal::ConnectEntry;
+using brave_vpn::internal::CreateEntry;
+using brave_vpn::internal::DisconnectEntry;
+using brave_vpn::internal::GetPhonebookPath;
+using brave_vpn::internal::PrintRasError;
+using brave_vpn::internal::RemoveEntry;
+
+namespace brave_vpn {
+
+// static
+BraveVPNConnectionManager* BraveVPNConnectionManager::GetInstance() {
+  static base::NoDestructor<BraveVPNConnectionManagerWin> s_manager;
+  return s_manager.get();
+}
+
+BraveVPNConnectionManagerWin::BraveVPNConnectionManagerWin() = default;
+BraveVPNConnectionManagerWin::~BraveVPNConnectionManagerWin() = default;
+
+void BraveVPNConnectionManagerWin::CreateVPNConnection(
+    const BraveVPNConnectionInfo& info) {
+  const std::wstring name = base::UTF8ToWide(info.name);
+  const std::wstring host = base::UTF8ToWide(info.url);
+  const std::wstring user = base::UTF8ToWide(info.id);
+  const std::wstring password = base::UTF8ToWide(info.pwd);
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&CreateEntry, name.c_str(), host.c_str(), user.c_str(),
+                     password.c_str()),
+      base::BindOnce(&BraveVPNConnectionManagerWin::OnCreated,
+                     weak_factory_.GetWeakPtr(), info.name));
+}
+
+void BraveVPNConnectionManagerWin::UpdateVPNConnection(
+    const BraveVPNConnectionInfo& info) {
+  NOTIMPLEMENTED();
```

Comment:
This will be an interesting case; I'm guessing a user may get a different server name / username / password if they choose a different location. In this case, we would want to support overwriting the existing connection so it has the new information

Reply:
yeah, maybe this update could be done by delete and add ?

Reply:
How about fill this api when we have concrete UX?

Reply:
SGTM! 😄